### PR TITLE
[FIX] mass_mailing: keep body of invalid saved mailing

### PR DESCRIPTION
**Pull request by abd-msyukyu-odoo**
There was an issue when a mailing save failed: the edited body of the mailing
was lost.

Impacted versions:

  - 15.0 and higher

Steps to reproduce:

  1. Create a mailing "Start from scratch" and add some elements
  2. Do not fill in the subject (or at least one required field)
  3. Save (it should fail because of the empty required field)

Current behavior:

  - The edited template is lost

Expected behavior:

  - Keep the edited template even if the save failed

This commit fixes the problem by properly awaiting the promise _doAction so that
the field's value will be set correctly.

The testing tour `mass_mailing_editor_tour` is updated to check that the
body_arch is kept after a failed save.

This issue was introduced in 15.0 with:
https://github.com/odoo/odoo/commit/95754eb6a464894417e8f763ae0fc4933ca3a157

Task-2856742

### CODE REVIEW

File: addons/mass_mailing/static/src/js/mass_mailing_widget.js [hunk 1]
```diff
@@ -61,9 +61,9 @@ var MassMailingFieldHtml = FieldHtml.extend({
 
         var $editable = this.wysiwyg.getEditable();
         await this.wysiwyg.cleanForSave();
-        return this.wysiwyg.saveModifiedImages(this.$content).then(function () {
+        return this.wysiwyg.saveModifiedImages(this.$content).then(async function () {
```
Comment (std-odoo) [inline addons/mass_mailing/static/src/js/mass_mailing_widget.js:64]: while you are at it, can you use arrow function ?
```suggestion
        return this.wysiwyg.saveModifiedImages(this.$content).then(async () => {
```

File: addons/mass_mailing/static/src/js/tours/mass_mailing_editor_tour.js [hunk 2]
```diff
@@ -10,16 +10,41 @@ odoo.define('mass_mailing.mass_mailing_editor_tour', function (require) {
...
         trigger: 'li.ui-menu-item',
     }, {
-        trigger: 'iframe #default',
-        run: () => $('iframe').contents().find('o_editable').html('<p><br/></p>'),
+        trigger: '[name="body_arch"] iframe #empty',
+    }, {
+        trigger: '[name="body_arch"] iframe .o_editable'
+    }, {
+        trigger: '[name="body_arch"] iframe #email_designer_default_body [name="Title"] .ui-draggable-handle',
+        run: function (actions) {
```
Comment (std-odoo) [inline addons/mass_mailing/static/src/js/tours/mass_mailing_editor_tour.js:23]: (same below)
```suggestion
        run: (actions) => {
```

File: addons/mass_mailing/static/src/js/tours/mass_mailing_editor_tour.js [hunk 3]
```diff
@@ -10,16 +10,41 @@ odoo.define('mass_mailing.mass_mailing_editor_tour', function (require) {
         trigger: '.o_app[data-menu-xmlid="mass_mailing.mass_mailing_menu_root"]',
     }, {
         trigger: 'button.o_list_button_add',
-    }, {
-        trigger: 'input[name="subject"]',
-        run: 'text Test',
```
Comment (std-odoo) [inline addons/mass_mailing/static/src/js/tours/mass_mailing_editor_tour.js:15]: why this change ?

**Review by std-odoo — COMMENTED** [review]
@abd-msyukyu-odoo Hi :)

Thank you for the fix, I added some comment, you can ping me once it's done :)

File: addons/mass_mailing/static/src/js/tours/mass_mailing_editor_tour.js (hunk 3 above)
Comment (abd-msyukyu-odoo) [reply addons/mass_mailing/static/src/js/tours/mass_mailing_editor_tour.js:15]: This step is moved at the end of the tour, because I want to do a "Save" with an empty required field (the subject), so that the fixed flow can be tested

**Review by abd-msyukyu-odoo — COMMENTED** [review]
Thanks for the review ;p

File: addons/mass_mailing/static/src/js/mass_mailing_widget.js [hunk 4]
```diff
@@ -61,23 +60,23 @@ var MassMailingFieldHtml = FieldHtml.extend({
 
         var $editable = this.wysiwyg.getEditable();
         await this.wysiwyg.cleanForSave();
-        return this.wysiwyg.saveModifiedImages(this.$content).then(function () {
-            self._isDirty = self.wysiwyg.isDirty();
-            self._doAction();
+        return this.wysiwyg.saveModifiedImages(this.$content).then(async () => {
```
Comment (awa-odoo) [inline addons/mass_mailing/static/src/js/mass_mailing_widget.js:63]: I think this comes from a validation comment so I'm sorry but I'l going to ask you the opposite, please revert to the "async function ()" code.

Why? Because this changes a lot of lines here below and it makes it hard to see the actual fix.
We usual don't do "linting" during bug fixes, to keep a minimal diff.

Also, it modifies the line with ".html", requiring us to do an security override for the mergebot, which could be avoided.

File: addons/mass_mailing/static/src/js/mass_mailing_widget.js [hunk 5]
```diff
@@ -61,23 +60,23 @@ var MassMailingFieldHtml = FieldHtml.extend({
...
+                dataPointID: this.dataPointID,
+                changes: _.object([fieldName], [this._unWrap($editable.html())])
             });
 
-            $editable.html(self.value);
-            if (self._isDirty && self.mode === 'edit') {
-                return self._doAction();
+            $editable.html(this.value);
+            if (this._isDirty && this.mode === 'edit') {
+                return this._doAction();
```
Comment (awa-odoo) [inline addons/mass_mailing/static/src/js/mass_mailing_widget.js:79]: This "doAction" can stay without an await?

File: addons/mass_mailing/static/src/js/tours/mass_mailing_editor_tour.js [hunk 6]
```diff
@@ -10,16 +10,41 @@ odoo.define('mass_mailing.mass_mailing_editor_tour', function (require) {
...
     }, {
         trigger: 'div[name="contact_list_ids"] .o_input_dropdown > input[type="text"]',
     }, {
         trigger: 'li.ui-menu-item',
     }, {
-        trigger: 'iframe #default',
-        run: () => $('iframe').contents().find('o_editable').html('<p><br/></p>'),
+        trigger: '[name="body_arch"] iframe #empty',
+    }, {
+        trigger: '[name="body_arch"] iframe .o_editable'
```
Comment (awa-odoo) [inline addons/mass_mailing/static/src/js/tours/mass_mailing_editor_tour.js:20]: Do we need to click twice on the iframe?

File: addons/mass_mailing/static/src/js/tours/mass_mailing_editor_tour.js [hunk 7]
```diff
@@ -10,16 +10,41 @@ odoo.define('mass_mailing.mass_mailing_editor_tour', function (require) {
...
+    }, {
+        trigger: '[name="body_arch"] iframe #email_designer_default_body [name="Title"] .ui-draggable-handle',
```
Comment (awa-odoo) [inline addons/mass_mailing/static/src/js/tours/mass_mailing_editor_tour.js:22]: Would you mind adding some explanation before your trigger to explain a bit what the step is about?

```js
// drag and drop a "body" snippet into the mail body editor
trigger: '[name="body_arch"] iframe #email_designer_default_body [name="Title"] .ui-draggable-handle',
```

(Not sure that's what it does, but it's to give an example).

Same for steps below unless they are crystal clear.

File: addons/mass_mailing/static/src/js/tours/mass_mailing_editor_tour.js [hunk 8]
```diff
@@ -10,16 +10,41 @@ odoo.define('mass_mailing.mass_mailing_editor_tour', function (require) {
...
+        trigger: '[name="body_arch"] iframe .odoo-editor-editable',
+        run: function () {
+            const ownerDocument = this.$anchor[0].ownerDocument;
+            const paragraph = ownerDocument.createElement('p');
+            const linebreak = ownerDocument.createElement('br');
+            paragraph.append(linebreak);
+            const fragment = new DocumentFragment();
+            fragment.append(paragraph);
+            this.$anchor[0].replaceChildren(...fragment.childNodes);
+        }
```
Comment (awa-odoo) [inline addons/mass_mailing/static/src/js/tours/mass_mailing_editor_tour.js:44]: Explanations would be especially needed here.
Seems like a lot of manual DOM manipulations, is it really required?

**Review by awa-odoo — COMMENTED** [review]
Hello @abd-msyukyu-odoo 

Thank you for the fix.
I added some comments / questions.

About your commit message, would you mind re-writing it a bit to match the style of what we usually do?
No need to break it down into bullet points I think.

> There was an issue with ...
> 
> This commit fixes the problem by...
> 
> This issue was introduced in: ...
> 
> [Links]

Thanks & cheers!

File: addons/mass_mailing/static/src/js/mass_mailing_widget.js (hunk 5 above)
Comment (abd-msyukyu-odoo) [reply addons/mass_mailing/static/src/js/mass_mailing_widget.js:79]: Yeah since the original signature returns a Promise, it is the responsibility of the caller of `commitChanges` to properly await if needed. The first _doAction need to be waited for because `this.value` is used afterwards assuming that the value was set.

File: addons/mass_mailing/static/src/js/tours/mass_mailing_editor_tour.js (hunk 6 above)
Comment (abd-msyukyu-odoo) [reply addons/mass_mailing/static/src/js/tours/mass_mailing_editor_tour.js:20]: The second click is indeed not needed, I'll put an empty function in `run` instead. (it is a verification step to check that the chosen configuration (#empty) finished rendering)

File: addons/mass_mailing/static/src/js/tours/mass_mailing_editor_tour.js (hunk 8 above)
Comment (abd-msyukyu-odoo) [reply addons/mass_mailing/static/src/js/tours/mass_mailing_editor_tour.js:44]: It is the verbose way to avoid using `.html("<p><br/></p>")` which would trigger the security flag, but maybe we don't care since it's a tour ?

And the justification for the step is that because we chose a "drag&drop" "template" to edit the mailing, the `body_arch` now contains a lot of preparatory nodes, and to avoid modifying the asserts in the python test which calls this tour, I replaced these nodes by the default value of `odoo-editor-editable` (an empty paragraph)

Comment (awa-odoo) [reply addons/mass_mailing/static/src/js/tours/mass_mailing_editor_tour.js]: Still don't think this is a good idea to do that many manual DOM manipulations.
This essentially forces tags into the editor instead of simulating user inputs (clicks, text, ...), which is what a tour is supposed to do.

Either you can easily "reset" the editor by typing something into it using a default "text Blabla" step.
Either you should create a second (small) tour that tests your specific use case.

Comment (abd-msyukyu-odoo) [reply addons/mass_mailing/static/src/js/tours/mass_mailing_editor_tour.js]: I'll remove this step and change the python asserts. They weren't particularly well chosen for the intended purpose anyway (prone to fail if the style is tweaked in the future)

File: addons/mass_mailing/tests/test_mailing_ui.py [hunk 9]
```diff
@@ -18,9 +18,12 @@ def test_01_mass_mailing_editor_tour(self):
         # The tour created and saved an email. The edited version should be
         # saved in body_arch, and its transpiled version (see convert_inline)
         # for email client compatibility should be saved in body_html. This
-        # ensures both fields have different values.
-        self.assertEqual(mail.body_arch, '<p><br></p>')
-        self.assertEqual(mail.body_html, '<p style="margin:0px 0 14px 0;box-sizing:border-box;"><br></p>')
+        # ensures both fields have different values (the mailing body should
+        # have been converted to a table in body_html).
+        self.assertTrue(mail.body_arch.find('data-snippet="s_title"'))
```
Comment (awa-odoo) [inline addons/mass_mailing/tests/test_mailing_ui.py:23]: Hello @Zinston 

Would you mind having a quick look at this modification?
It seems to still do the appropriate check, but I prefer to be sure.

Thanks!

Comment (Zinston) [reply addons/mass_mailing/tests/test_mailing_ui.py:23]: Hi @awa-odoo ! Looks good to me, this is gonna be more robust btw, thanks for the change ;-)

**Comment by Zinston** [comment]
@abd-msyukyu-odoo Are you force pushing _after_ @awa-odoo started the merge??

**Comment by abd-msyukyu-odoo** [comment]
@Zinston Yeah I failed my checks on Markup.find, the "not found" value is -1, not 0 or False :/ I hope the runbot will abort the operation

**Comment by Zinston** [comment]
So I guess it was too late ? :eyes:

**Comment by awa-odoo** [comment]
> So I guess it was either too late or it did indeed abort already automatically ? :eyes: 

@Zinston once you force-push it automatically removes the "review" tag on the runbot, you can check that on the dashboard status (very first message posted by the bot on the PR).

It has then to be r+ed again (maybe you already knew this).

In other words, we're good ^^

**Comment by Zinston** [comment]
@awa-odoo Great. Well, better safe than sorry ;-)

**Comment by tde-banana-odoo** [comment]
@robodoo r+ then
